liquid-rust-syntax/src/surface.rs

```diff
@@ -1,60 +1,86 @@
 pub use rustc_ast::token::LitKind;
+use rustc_ast::Mutability;
+use rustc_hir::def_id::DefId;
+use rustc_middle::ty;
+pub use rustc_middle::ty::{FloatTy, IntTy, ParamTy, UintTy};
 pub use rustc_span::symbol::Ident;
 use rustc_span::Span;
 
 use crate::ast::{self, Expr, GenericParam, Refine};
 
 #[derive(Debug)]
-pub struct FnSig {
+pub enum BareSig {
+    AstSig(ast::FnSig),
+    SurSig(BareFnSig),
+}
+
+#[derive(Debug)]
+pub struct FnSig<T> {
     /// example: `l: i32@n`
-    pub requires: Vec<(Ident, Ty)>,
+    pub requires: Vec<(Ident, Ty<T>)>,
     /// example `i32{v:v >= 0}`
-    pub returns: Ty,
+    pub returns: Ty<T>,
     /// example: `*x: i32{v:v = n+1}`
-    pub ensures: Vec<(Ident, Ty)>,
+    pub ensures: Vec<(Ident, Ty<T>)>,
     /// example: `where n > 0`
     pub wherep: Option<Expr>,
     /// source span
     pub span: Span,
 }
 
 #[derive(Debug)]
-pub struct Ty {
-    pub kind: TyKind,
+pub struct Ty<T> {
+    pub kind: TyKind<T>,
     pub span: Span,
 }
 
 #[derive(Debug)]
-pub enum TyKind {
+pub enum TyKind<T> {
     /// ty
-    Base(Path),
+    Base(Path<T>),
+
+    /// type parameters (not used in `Bare` but will show up after zipping)
+    Param(ty::ParamTy),
```

Comment:
TODO

This variant is redundant. It should be covered by  `DefTyKind::Base(..)`.